### ISSUE TITLE
Fn doesn't handle CPU extensive consumption container termination when by Docker

### ISSUE THREAD

**Issue by denismakogon** [fnproject/fn#1293]
**Description**

There two exceptional cases when Docker terminates the container:
 - OOM
 - CPU extensive usage

It seems like we've battled docker OOM, but CPU extensive usage problem is still there.
When a function attempts to consume 100% of its CPU quota Docker will kill it immediately.

Unfortunately, there are no signs of this exceptional case in Fn logs, all you get is:
```
time="2018-11-01T08:52:07Z" level=error msg="Got error from UDS socket" action="server.handleHTTPTriggerCall)-fm" appName=ffmpeg app_id=01CV3BC2QANG8G00RZJ0000001 error="Post http://localhost/call: EOF" fn_id=01CV55XW0NNG8G00GZJ0000003 id=01CV77HA6ZNG8G00GZJ000000A triggerSource=/timecode-statistics
time="2018-11-01T08:52:07Z" level=info msg="hot function terminating" app_id=01CV3BC2QANG8G00RZJ0000001 cpus= error="error receiving function response" fn_id=01CV55XW0NNG8G00GZJ0000003 id=01CV77HA6ZNG8G00GZJ000000C idle_timeout=30 image="denismakogon/timecode-statistics:0.0.2" memory=1024
```

**Steps to reproduce the issue:**

Here i can provide all the necessary code and artifacts to test the function, just ping me in slack.

**Describe the results you received:**

Function dies with no signs of a problem.
In Fn logs i see only that Fn was not able to read the response from a function.

**Describe the results you expected:**

Something more meaningful, like, your function is out, it's dead because not enough CPU/RAM. Would be nice to have 2 different error codes for this: one for CPU, one for RAM.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `fn version` (CLI command):**

```
Client version: 0.5.23 is not latest: 0.5.26
Server version:  0.3.604
```

**Additional environment details (OSX, Linux, flags, etc.):**

Mac OS Mojave. Docker-4-Mac.

**Comment by zootalures** [comment]
I do not think that docker kills a container when it consumes high CPU, can you check your function and docker setup

**Comment by rdallman** [comment]
this issue lacks a smoking gun, it could be anything... turn on debug logs and get a stack trace on program exit or something, please. it could be even your OS doing this (though I've never heard of such a thing for CPU in an init system or similar, it could easily be scripted), a clean-room reproduction script for this issue may help debug this. we're also not trying to plumb out container level issues to function output, we're preferring to put this in logging or metrics as for end consumers of functions (say, hitting a web url and getting back html) this doesn't really make sense.

**Comment by denismakogon** [comment]
It turns out that docker really killed my function due to consisten 100% CPU usage. In Docker logs it looked like OOM, basically RAM exceeded, but function has never reached 70% of its RAM allocated. But according to docker metrics RAM consumption has never been up to event 46-54% of 1.5GB. However, I added more RAM up to 2Gb and (average consumption was 1375Mb) and my code worked because it got more CPU quota from Fn.

**Comment by rdallman** [comment]
so, we talked about this a bit, but posting for posterity, but actual ram usage isn't indicative of whether your process is safe from the oom killer. it's a complex topic that maybe is explained by someone more thoughtful and intelligent than I: http://wiki.dreamrunner.org/public_html/Linux/Linux-Memory.html

in any case, this smells like it's oom killer pretty bad.

**Comment by denismakogon** [comment]
Okay, so, OOM problem is there and we can't change the way memory model works.
But here's another problem, from a user perspective: a function that dies during the sync call should be somehow handled, at least to report to a user that a function needs more RAM to process, but not something like "hey, can't get the response, have no idea why" (there are few more cases when the response messages are not quite accurate).

**Comment by rdallman** [comment]
@denismakogon there should be a log message? there's also a metric, granted metrics is kind of an open ended issue in general. we're hoping to have all stuff like this around container/function failures themselves in the logs and metrics for users, not on the function output. does that make sense to you or is there additional something you'd want as both function consumer and function dev?
